### PR TITLE
Implement a common scheduler

### PR DESCRIPTION
**Pull request by ywk253100**
Implement a common scheduler that can be used globally

Signed-off-by: Wenkai Yin <yinw@vmware.com>

### CODE REVIEW

**Comment by steven-zou** [comment]
@ywk253100 

So you'll add UT in a new PR? Right?

File: src/core/router.go [hunk 1]
```diff
@@ -131,6 +132,7 @@ func initRouters() {
 	beego.Router("/service/notifications/jobs/adminjob/:id([0-9]+)", &admin.Handler{}, "post:HandleAdminJob")
 	beego.Router("/service/notifications/jobs/replication/:id([0-9]+)", &jobs.Handler{}, "post:HandleReplicationScheduleJob")
 	beego.Router("/service/notifications/jobs/replication/task/:id([0-9]+)", &jobs.Handler{}, "post:HandleReplicationTask")
+	beego.Router("/service/notifications/scheduler", &scheduler.Handler{}, "post:Handle")
```
Comment (steven-zou) [inline src/core/router.go]: Is this the hook endpoint?



----

Better than `/api/schedules`?

POST `/api/schedules` for schedule 
DELETE `/api/schedules/:sid` for unschedule?

File: src/core/service/notifications/scheduler/handler.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+			h.SendInternalServerError(fmt.Errorf("failed to update status of job %s: %v", data.JobID, err))
+			return
+		}
+		return
+	}
+
+	// run callback function
+	// just log the error message when handling check in request if got any error
+	params := map[string]interface{}{}
+	if err := json.Unmarshal([]byte(data.CheckIn), &params); err != nil {
```
Comment (steven-zou) [inline src/core/service/notifications/scheduler/handler.go]: backoff is directly saving it as pure text

File: src/pkg/scheduler/model/schedule.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+package model
+
+import (
+	"time"
+)
+
+// Schedule is a record for a scheduler job
+type Schedule struct {
+	ID            int64      `orm:"pk;auto;column(id)" json:"id"`
+	ReferenceType string     `orm:"column(reference_type)" json:"reference_type"` // the type of reference that the schedule created for, e.g. retention
```
Comment (steven-zou) [inline src/pkg/scheduler/model/schedule.go]: To?

```
struct Reference {
  Type string
  ID string 
}
```

File: src/pkg/scheduler/periodic_job.go [hunk 4]
```diff
@@ -40,5 +46,9 @@ func (pj *PeriodicJob) Validate(params job.Parameters) error {
 
 // Run the job
 func (pj *PeriodicJob) Run(ctx job.Context, params job.Parameters) error {
-	return ctx.Checkin(fmt.Sprintf("pong=%d", time.Now().Unix()))
+	data, err := json.Marshal(params)
```
Comment (steven-zou) [inline src/pkg/scheduler/periodic_job.go]: So, the `checkin` will be definitely structured data?

File: src/pkg/scheduler/scheduler.go [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+type scheduler struct {
+	sync.RWMutex
+	internalCoreURL  string
+	registry         map[string]CallbackFunc
+	manager          Manager
+	jobserviceClient cjob.Client
+}
+
+func (s *scheduler) Schedule(referenceType string, referenceID int64,
```
Comment (steven-zou) [inline src/pkg/scheduler/scheduler.go]: Need more discussions here, especially the arguments and returns.

File: src/pkg/scheduler/scheduler.go [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+		StatusHook: statusHookURL,
+	}
+	jobID, err := s.jobserviceClient.SubmitJob(jd)
+	if err != nil {
+		return 0, err
+	}
+	log.Debugf("the scheduler job submitted to Jobservice, job ID: %s", jobID)
+
+	// create schedule record
+	now := time.Now()
```
Comment (steven-zou) [inline src/pkg/scheduler/scheduler.go]: Can we create record first and then submit the job? If we failed to submit the job, then mark the status to error.

File: src/pkg/scheduler/scheduler.go [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+	return f, nil
+}
+
+func callbackFuncExist(name string) bool {
+	_, exist := registry[name]
+	return exist
+}
+
+// New returns an instance of the default scheduler
+func New() Scheduler {
```
Comment (steven-zou) [inline src/pkg/scheduler/scheduler.go]: Provide a `scheduler.DefaultScheduler` here for other packages to call?

File: src/pkg/scheduler/handler.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+		manager: NewManager(),
+	}
+}
+
+type handler struct {
+	manager Manager
+}
+
+func (h *handler) UpdateStatus(jobID, status string) error {
+	schedules, err := h.manager.List(&model.ScheduleQuery{
```
[the remark this answers is not in the document]
Comment (ywk253100) [reply src/pkg/scheduler/handler.go]: Query by job ID, not the task ID

**Pull request merged**